File: cli/internal/connectapi/reactions.go
package connectapi

import (
	"context"

	"connectrpc.com/connect"
	"hmans.de/chatto/internal/core"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
)

func (s *messageService) AddReaction(ctx context.Context, req *connect.Request[apiv1.AddReactionRequest]) (*connect.Response[apiv1.AddReactionResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	added, err := s.api.core.ReactionModel().AddReaction(ctx, core.ReactionMutationInput{
		ActorID:        caller.UserID,
		RoomID:         req.Msg.RoomId,
		MessageEventID: req.Msg.MessageEventId,
		Emoji:          req.Msg.Emoji,
	})
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.AddReactionResponse{
		Added:    added,
		Reaction: s.reactionSummary(ctx, caller.UserID, req.Msg.MessageEventId, req.Msg.Emoji),
	}), nil
}

func (s *messageService) RemoveReaction(ctx context.Context, req *connect.Request[apiv1.RemoveReactionRequest]) (*connect.Response[apiv1.RemoveReactionResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	removed, err := s.api.core.ReactionModel().RemoveReaction(ctx, core.ReactionMutationInput{
		ActorID:        caller.UserID,
		RoomID:         req.Msg.RoomId,
		MessageEventID: req.Msg.MessageEventId,
		Emoji:          req.Msg.Emoji,
	})
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.RemoveReactionResponse{
		Removed:  removed,
		Reaction: s.reactionSummary(ctx, caller.UserID, req.Msg.MessageEventId, req.Msg.Emoji),
	}), nil
}

func (s *messageService) reactionSummary(ctx context.Context, viewerID, messageEventID, emoji string) *apiv1.RoomTimelineReaction {
	summaries, err := s.api.core.GetReactions(ctx, messageEventID)
	if err != nil {
		return nil
	}
	for _, summary := range summaries {
		if summary.Emoji != emoji {
			continue
		}
		userIDs := firstN(summary.UserIDs, 5)
		return &apiv1.RoomTimelineReaction{
			Emoji:          summary.Emoji,
			Count:          int32(len(summary.UserIDs)),
			HasReacted:     containsString(summary.UserIDs, viewerID),
			PreviewUserIds: userIDs,
		}
	}
	return nil
}

File: apps/frontend/src/lib/api-client/reactions.ts
import { authHeaders, createChattoClient, handleAuthError } from "./connect.js";
import { MessageService } from "@chatto/api-types/api/v1/messages_connect";
import type { RoomTimelineReaction } from "@chatto/api-types/api/v1/room_timeline_pb";

export type ConnectAPIConfig = {
  serverId?: string;
  baseUrl: string;
  bearerToken: string | null;
  onAuthenticationRequired?: (serverId: string) => void;
};

export type ReactionInput = {
  roomId: string;
  messageEventId: string;
  emoji: string;
};

export type ReactionSummary = {
  emoji: string;
  count: number;
  hasReacted: boolean;
  previewUserIds: string[];
};

export type AddReactionResult = {
  added: boolean;
  reaction: ReactionSummary | null;
};

export type RemoveReactionResult = {
  removed: boolean;
  reaction: ReactionSummary | null;
};

export function createReactionAPI(config: ConnectAPIConfig) {
  const client = createChattoClient(MessageService, config);
  const headers = () => authHeaders(config);
  return {
    async addReaction(input: ReactionInput): Promise<AddReactionResult> {
      try {
        const response = await client.addReaction(input, {
          headers: headers(),
        });
        return {
          added: response.added,
          reaction: mapReactionSummary(response.reaction),
        };
      } catch (err) {
        return handleAuthError(config, err);
      }
    },

    async removeReaction(input: ReactionInput): Promise<RemoveReactionResult> {
      try {
        const response = await client.removeReaction(input, {
          headers: headers(),
        });
        return {
          removed: response.removed,
          reaction: mapReactionSummary(response.reaction),
        };
      } catch (err) {
        return handleAuthError(config, err);
      }
    },
  };
}

function mapReactionSummary(
  reaction: RoomTimelineReaction | undefined,
): ReactionSummary | null {
  if (!reaction || !reaction.emoji) return null;
  return {
    emoji: reaction.emoji,
    count: reaction.count,
    hasReacted: reaction.hasReacted,
    previewUserIds: [...reaction.previewUserIds],
  };
}
